app/src/main/java/io/homeassistant/companion/android/widgets/ButtonWidgetConfigureActivity.kt

```diff
@@ -38,72 +48,116 @@ class ButtonWidgetConfigureActivity : Activity() {
+            service
         )
+
+        // Fetch and send label and icon
         intent.putExtra(
             ButtonWidget.EXTRA_LABEL,
-            label.text.toString()
+            context.label.text.toString()
         )
         intent.putExtra(
             ButtonWidget.EXTRA_ICON,
             context.widget_config_spinner.selectedItemId.toInt()
         )
-        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+
+        // Analyze and send service data
+        val serviceDataMap = HashMap<String, Any>()
+        for (i in 0 until dynamicFields.size) {
+            val fieldText = dynamicFieldAdapter.getText(i)
+
+            // Don't store data that's empty (or just whitespace)
+            if (!fieldText.isBlank()) {
+                serviceDataMap[dynamicFields[i].second] = fieldText
+            }
+        }
+
+        intent.putExtra(
+            ButtonWidget.EXTRA_SERVICE_DATA,
+            Gson().toJson(serviceDataMap)
+        )
+
         context.sendBroadcast(intent)
 
         // Make sure we pass back the original appWidgetId
         setResult(RESULT_OK, Intent().putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId))
         finish()
     }
 
+    private val dropDownOnFocus = View.OnFocusChangeListener { view, hasFocus ->
+        if (hasFocus && view is AutoCompleteTextView) {
+            view.showDropDown()
+        }
+    }
+
+    private val serviceTextWatcher: TextWatcher = (object : TextWatcher {
+        override fun afterTextChanged(p0: Editable?) {
+            val serviceText: String = p0.toString()
+
+            if (services.keys.contains(serviceText)) {
+                Log.d(TAG, "Valid domain and service--processing dynamic fields")
+
+                // Make sure there are not already any dynamic fields created
+                // This can happen if selecting the drop-down twice or pasting
+                widget_config_fields_layout.removeAllViews()
+
+                // We only call this if servicesAvailable was fetched and is not null,
+                // so we can safely assume that it is not null here
+                val fields = services[serviceText]!!.serviceData.fields
+                val fieldKeys = fields.keys
+                Log.d(TAG, "Fields applicable to this service: $fields")
+
+                fieldKeys.sorted().forEach { fieldKey ->
+                    Log.d(TAG, "Creating a text input box for $fieldKey")
+
+                    // Insert a dynamic layout
+                    // IDs get priority and go at the top, since the other fields
+                    // are usually optional but the ID is required
+                    if (fieldKey.contains("_id"))
+                        dynamicFields.add(0, Pair(serviceText, fieldKey))
+                    else
+                        dynamicFields.add(Pair(serviceText, fieldKey))
+
+                    dynamicFieldAdapter.notifyDataSetChanged()
```

Comment:
Don't need to call this every time we update the fields, once at the end should be plenty.

Reply:
Oh, whoops, thought I put that outside the for loop. I agree.

Reply:
Corrected in commit 29fc233